Product/ProductManager.php

```diff
@@ -803,8 +803,45 @@ public function save(
             }
         }
 
-        // TODO: handle attributes
-
+        if (isset($data['attributes'])) {
+            // Delete attributes
+            $attributeIds = [];
+            foreach ($data['attributes'] as $attributeData) {
+                if (isset($attributeData['attributeId'])) {
+                    $attributeIds[] = $attributeData['attributeId'];
+                }
+            }
+            // If attributes is not in current specified attributes remove it from product
+            foreach ($product->getAttributes() as $productAttribute) {
+                if (!in_array($productAttribute->getAttribute()->getId(), $attributeIds)) {
+                    $product->getEntity()->removeProductAttribute($productAttribute->getEntity());
+                    $this->em->remove($productAttribute->getEntity());
+                }
+            }
+            // Add and change attributes
+            foreach ($data['attributes'] as $attributeData) {
+                $attributeValue = $attributeData['value'];
+                // Check if attribute exists for current product
+                $productAttribute = $this->productAttributeRepository->findByAttributeIdAndProductId(
```

Comment:
actually you just iterated over all product-attributes, so this query is not really necessary (i'm not sure oh how good doctrine caches these)